cmd/graph.go

```diff
@@ -23,22 +23,67 @@ import (
 	"github.com/spf13/cobra"
 )
 
+var dep string
+
 var graphCmd = &cobra.Command{
 	Use:   "graph",
 	Short: "Generate a .dot file to be used with Graphviz's dot command.",
 	Long: `A graph.dot file will be generated which can be used with Graphviz's dot command.
 	For example to generate a svg image use:
 	twopi -Tsvg -o dag.svg graph.dot`,
 	RunE: func(cmd *cobra.Command, args []string) error {
-		depGraph, deps, _ := getDepInfo()
-		fileContents := "digraph {\ngraph [rankdir=TB, overlap=false];\n"
-		for _, dep := range deps {
-			_, ok := depGraph[dep]
-			if !ok {
-				continue
+		depGraph, deps, mainModule := getDepInfo()
+		// strict ensures that there is only one edge between two vertices
+		// overlap = false ensures the vertices don't overlap
+		fileContents := "strict digraph {\ngraph [overlap=false];\n"
+
+		// graph to be generated is based around input dep
+		if dep != "" {
+			var chains []Chain
+			var temp Chain
+			getAllChains(mainModule, depGraph, temp, &chains)
+			// to color the entered node as yellow
+			fileContents += fmt.Sprintf("MainNode [label=\"%s\", style=\"filled\" color=\"yellow\"]", dep)
+
+			// add all chains which have the input dep to the .dot file
+			for _, chain := range chains {
+				if chainContains(chain, dep) {
+					for i, d := range chain {
+						if i == len(chain)-1 {
+							if d == dep {
+								// for the input dep use a colored node
+								fileContents += "MainNode\n"
+							} else {
+								fileContents += fmt.Sprintf("\"%s\"\n", d)
+							}
+						} else {
+							if d == dep {
+								// for the input dep use a colored node
+								fileContents += "MainNode -> "
+							} else {
+								fileContents += fmt.Sprintf("\"%s\" -> ", d)
+							}
+						}
+					}
+				}
 			}
-			for _, neighbour := range depGraph[dep] {
-				fileContents += fmt.Sprintf("\"%s\" -> \"%s\"\n", dep, neighbour)
+		} else {
+			// color the main module as yellow
+			fileContents += fmt.Sprintf("MainNode [label=\"%s\", style=\"filled\" color=\"yellow\"]", mainModule)
+			for _, dep := range deps {
+				_, ok := depGraph[dep]
+				if !ok {
+					continue
+				}
+				// main module can never be a neighbour
+				for _, neighbour := range depGraph[dep] {
+					if dep == mainModule {
```

Comment:
potentially you could extract the nodes creation to func and reuse it here as well as in the other branch

Reply:
I have switched to `strings.Join()` for the `if` block but in the `else` block, reusing the code won't be possible IMO. 

This is because for generating the graph based on an input dependency our `.dot` file looks like (the `if` block):
```
"A" -> "B" -> "C" -> "D"
"E" -> "F" -> "G" -> "H"
.
.
.
```

but for generating the graph of the entire dependency tree it looks like(the `else` block):
```
"A" -> "B"
"B" -> "C"
"C" -> "D"
"E" -> "F"
"F" -> "G"
"G" -> "H"
```

Now we could have the second graph also look like the first one but for that we would need access to all the chains like we have incase of the first graph. Getting these chains using the `getAllChains` function is both time and memory intensive so it'd be better to avoid it. Thoughts?

Reply:
Hmm I see, I think I am lacking an understanding what's the difference between the two branches 🤔  Why is it different?

Reply:
So we support making two types of graphs:

- based on a dependency a user enters
- graph of all dependencies

To create the graph of all dependencies [the logic](https://github.com/kubernetes-sigs/depstat/pull/36/files#diff-171d9186355919e81f4548c3a10b7500b2702557a9ac818bf2cd820124358d89R68-R74) is simple. We iterate through each dependency and then iterate through its neighbors and have all the edges of the graph in the `.dot` file like this:

```
"A" -> "B"
"B" -> "C"
"C" -> "D"
"E" -> "F"
"F" -> "G"
"G" -> "H"
```


But for the graph based on the dependency entered by the user, we need access to **all dependency chains**. This is because we need to check which chain has the dependency entered by the user. Once we know that we simply put those chains in the `.dot` file:

```
"A" -> "B" -> "C" -> "D"
"E" -> "F" -> "G" -> "H"
```


**1. Why can't we reuse the logic by abstracting it to a function?**

Reusing the logic would mean having the graph of all dependencies in a format similar to the graph based on the dependency entered by the user.

And to create this graph in a similar format, we would need access to **all dependency chains**. 

**2. What's the problem in getting access to all dependency chains?**

To get these chains we call the `getAllChains()` [function](https://github.com/kubernetes-sigs/depstat/pull/36/files#diff-171d9186355919e81f4548c3a10b7500b2702557a9ac818bf2cd820124358d89R45). This function is very demanding in terms of memory and also takes a lot of time, so we should ideally try to avoid calling it. Since we can get the graph of all dependencies without calling this function I didn't use it.



Does this make more sense? 😅 Also, I'm currently refactoring the code in this file into separate functions and will push that tomorrow, which would help seeing the difference a bit more clearly.